contrib/gorm.io/gorm/gorm.go

```diff
@@ -0,0 +1,168 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package gorm provides helper functions for tracing the gorm.io/gorm package (https://github.com/go-gorm/gorm).
+package gorm
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"gorm.io/gorm"
+)
+
+type key string
+
+const (
+	gormConfigKey        = key("dd-trace-go:config")
+	gormSpanStartTimeKey = key("dd-trace-go:span")
+)
+
+// Open opens a new (traced) database connection. The used driver must be formerly registered
+// using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
+func Open(dialector gorm.Dialector, cfg *gorm.Config, opts ...Option) (*gorm.DB, error) {
+	db, err := gorm.Open(dialector, cfg)
+	if err != nil {
+		return db, err
+	}
+
+	return WithCallbacks(db, opts...)
+}
+
+// WithCallbacks registers callbacks to the gorm.DB for tracing.
+// It should be called once, after opening the db.
+// The callbacks are triggered by Create, Update, Delete,
+// Query and RowQuery operations.
+func WithCallbacks(db *gorm.DB, opts ...Option) (*gorm.DB, error) {
+	afterFunc := func(operationName string) func(*gorm.DB) {
+		return func(db *gorm.DB) {
+			after(db, operationName)
+		}
+	}
+
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+
+	ctx := context.Background()
+	if db.Statement != nil && db.Statement.Context != nil {
+		ctx = db.Statement.Context
+	}
+
+	db = db.WithContext(context.WithValue(ctx, gormConfigKey, cfg))
```

Comment:
Rather than putting `cfg` into the context, I think it is cleaner to just capture it in `afterFunc` and pass it as an argument to `after`. Then we can also avoid the config shuffle in `WithContext`.